// neuroninjas/index.js
const http = require('http');
const fs = require('fs');
const path = require('path');
function neuroninjas(){
    this.routes = [];
    this.middlewares = [];
    this.rateLimitConfig = null;
    this.ipStore = {};
    this.staticFolder = null;
}
neuroninjas.prototype.static = function(folderName){
    this.staticFolder = path.join(folderName);
};
neuroninjas.prototype.get = function(path, handler){
    this.routes.push({
        method: 'GET',
        path: path,
        handler: handler
    });
};
neuroninjas.prototype.post = function(path, handler){
    this.routes.push({
        method: 'POST',
        path: path,
        handler: handler
    });
};
neuroninjas.prototype.use = function(middleware){
    this.middlewares.push(middleware);
};
neuroninjas.prototype.rateLimit = function(options){
    this.rateLimitConfig = options;
};
neuroninjas.prototype.listen = function(port, callback){
    const server = http.createServer((req, res) => {
if(this.staticFolder){
    const filePath = path.join(this.staticFolder, req.url);
    if(filePath.startsWith(this.staticFolder)){ 
        fs.stat(filePath, (err, stats)=>{
            if(!err && stats.isFile()){
                const ext = path.extname(filePath);
                const contentTypes = {
                    ".html":"text/html",
                    ".css":"text/css",
                    ".js":"application/javascript",
                    ".json":"application/json",
                    ".png":"image/png",
                    ".jpg":"image/jpeg",
                    ".jpeg":"image/jpeg",
                    ".gif":"image/gif",
                    ".svg":"image/svg+xml"
                };
                res.writeHead(200, {
                    "Content-Type": contentTypes[ext] || "application/octet-stream"
                });
                fs.createReadStream(filePath).pipe(res);
            }
        });
    }
}
        if(this.rateLimitConfig){
            const ip = req.socket.remoteAddress;
            const now = Date.now();
            const windowTime = this.rateLimitConfig.window;
            const maxRequests = this.rateLimitConfig.max;
            if(!this.ipStore[ip]){
                this.ipStore[ip] = [];
            }
            this.ipStore[ip] = this.ipStore[ip].filter(
                timestamp => now - timestamp < windowTime
            );
            if(this.ipStore[ip].length >= maxRequests){
                res.writeHead(429, {'Content-Type':'text/plain'});
                return res.end("429 Too Many Requests");
            }
            this.ipStore[ip].push(now);
        }
        const route = this.routes.find(
            r => r.method === req.method && r.path === req.url
        );
        let i = 0;
        const next = () => {
            if(i < this.middlewares.length){
                this.middlewares[i++](req, new response(res, this.staticFolder), next);
            }
            else{
                if(route){
                    return route.handler(req, new response(res, this.staticFolder));
                }
                const dynamicRoute = this.routes.find(
                    r => r.method === req.method && r.path.includes(':')
                );
                if(dynamicRoute){
                    const urlParts = req.url.split('/');
                    const routeParts = dynamicRoute.path.split('/');
                    if(urlParts.length === routeParts.length){
                        let params = {};
                        let matched = true;
                        for(let j=0; j<routeParts.length; j++){
                            if(routeParts[j].startsWith(':')){
                                const key = routeParts[j].slice(1);
                                params[key] = urlParts[j];
                            }
                            else if(routeParts[j] !== urlParts[j]){
                                matched = false;
                                break;
                            }
                        }

                        if(matched){
                            req.params = params;
                            return dynamicRoute.handler(req, new response(res, this.staticFolder));
                        }
                    }
                }
                res.writeHead(404);
                res.end("Not Found");
            }
        };
        next();
    });
    server.listen(port, callback);
};
function response(res,staticFolder){
    this.res = res;
    this.staticFolder = staticFolder;
}
response.prototype.send = function(data){
    this.res.end(data);
};
response.prototype.json = function(data){
    this.res.setHeader('Content-Type','application/json');
    this.res.end(JSON.stringify(data));
};
response.prototype.render = function(view){
    const filename = path.join(this.staticFolder, view + '.html');
    console.log(filename)
    fs.readFile(filename, (err, data)=>{
        if(err){
            this.res.writeHead(500);
            this.res.end("Internal Server Error");
        }
        else{
            this.res.setHeader('Content-Type','text/html');
            this.res.end(data);
        }
    });
};
module.exports = neuroninjas;
